samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -69,87 +78,110 @@
    */
   private int nextOpNum = 0;
   private final Set<String> operatorIds = new HashSet<>();
-  private Serde<?> defaultSerde = new KVSerde(new NoOpSerde(), new NoOpSerde());
   private ContextManager contextManager = null;
+  private SystemDescriptor defaultSystemDescriptor;
 
   public StreamGraphSpec(Config config) {
     this.config = config;
   }
 
   @Override
-  public void setDefaultSerde(Serde<?> serde) {
-    Preconditions.checkNotNull(serde, "Default serde must not be null");
+  public void setDefaultSystem(SystemDescriptor<?, ?> defaultSystemDescriptor) {
+    String defaultSystemName = defaultSystemDescriptor.getSystemName();
     Preconditions.checkState(inputOperators.isEmpty() && outputStreams.isEmpty(),
-        "Default serde must be set before creating any input or output streams.");
-    this.defaultSerde = serde;
+        "Default system must be set before creating any input or output streams.");
+    Preconditions.checkState(!systemDescriptors.containsKey(defaultSystemName) ||
+            systemDescriptors.get(defaultSystemName).equals(defaultSystemDescriptor),
+        "Must not use different system descriptor instances for the same system name: " + defaultSystemName);
+    systemDescriptors.put(defaultSystemName, defaultSystemDescriptor);
+    this.defaultSystemDescriptor = defaultSystemDescriptor;
   }
 
   @Override
-  public <M> MessageStream<M> getInputStream(String streamId, Serde<M> serde) {
+  public <M> MessageStream<M> getInputStream(InputDescriptor<M, ?> inputDescriptor) {
+    Optional<SystemDescriptor> systemDescriptor = inputDescriptor.getSystemDescriptor();
+    if (systemDescriptor.isPresent() && systemDescriptor.get() instanceof ExpandingSystemDescriptor) {
+      return ((ExpandingSystemDescriptor) systemDescriptor.get()).getExpander().apply(this, inputDescriptor);
+    }
+
+    String streamId = inputDescriptor.getStreamId();
+    Serde serde = inputDescriptor.getSerde();
+
     Preconditions.checkState(isValidId(streamId),
         "streamId must be non-empty and must not contain spaces or special characters: " + streamId);
     Preconditions.checkNotNull(serde, "serde must not be null for an input stream.");
     Preconditions.checkState(!inputOperators.containsKey(streamId),
         "getInputStream must not be called multiple times with the same streamId: " + streamId);
+    Preconditions.checkState(!inputDescriptors.containsKey(streamId),
+        "getInputStream must not be called multiple times with the same input descriptor: " + streamId);
+    systemDescriptor.ifPresent(sd -> {
+        String systemName = sd.getSystemName();
+        Preconditions.checkState(!systemDescriptors.containsKey(systemName) || systemDescriptors.get(systemName).equals(sd),
```

Comment:
That makes sense. I guess I just saw the "equals" call with a missing hashcode/equals impl. Maybe it would help to use "==" explicitly or put a comment to clarify.